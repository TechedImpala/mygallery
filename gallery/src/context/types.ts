import { ApplicationStateEnum } from "../types";

export interface IAppProviderState {
  activeState: ApplicationStateEnum;
  setActiveState: React.Dispatch<
    React.SetStateAction<ApplicationStateEnum>
  > | null;
  activeTopic: string;
  setActiveTopic: React.Dispatch<React.SetStateAction<string>> | null;
  menuVisible: boolean;
  setMenuVisibility: React.Dispatch<React.SetStateAction<boolean>> | null;
}
