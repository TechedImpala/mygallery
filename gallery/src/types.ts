export enum ApplicationStateEnum {
  GRID = "GRID",
  MENU = "MENU",
}

export interface IUrls {
  raw: string;
  full: string;
  regular: string;
  small: string;
  thumb: string;
  small_s3: string;
}
