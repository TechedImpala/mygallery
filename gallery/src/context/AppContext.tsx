import { createContext, useContext, useState } from "react";
import { ApplicationStateEnum } from "../types";
import { IAppProviderState } from "./types";

const AppContext = createContext<IAppProviderState>({
  activeState: ApplicationStateEnum.MENU,
  setActiveState: null,
  activeTopic: "",
  setActiveTopic: null,
  menuVisible: true,
  setMenuVisibility: null,
});

interface IAppProviderProps {
  children: React.ReactNode | React.ReactNode[];
}

const AppProvider = ({ children }: IAppProviderProps) => {
  const [menuVisible, setMenuVisibility] = useState(true);
  const [activeState, setActiveState] = useState<ApplicationStateEnum>(
    ApplicationStateEnum.MENU
  );

  const [activeTopic, setActiveTopic] = useState<string>("");

  const value = {
    activeState,
    setActiveState,
    activeTopic,
    setActiveTopic,
    menuVisible,
    setMenuVisibility,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};

export const useAppContext = () => useContext(AppContext);

export default AppProvider;
