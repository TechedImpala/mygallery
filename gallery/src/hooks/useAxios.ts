import { useState, useEffect, useCallback } from "react";
import axios, { AxiosRequestConfig, AxiosResponse } from "axios";

const baseUrl = process.env.API_BASE_URL || "https://api.unsplash.com";

export const axiosInstance = axios.create({
  baseURL: baseUrl,
  responseType: "json",
});

interface IUseAxiosProps extends AxiosRequestConfig {
  manual?: boolean;
}

const useAxios = <T>(requestConfig: IUseAxiosProps) => {
  const [data, setData] = useState<T | null>(null);
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<unknown>(null);

  const { url } = requestConfig;

  const handleRequest = useCallback(() => {
    setLoading(true);
    axiosInstance
      .request(requestConfig)
      .then((response: AxiosResponse) => {
        setData(response.data);
        setLoading(false);
      })
      .catch((error: unknown) => {
        setError(error);
        setLoading(false);
      });
  }, [url]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (!requestConfig.manual) handleRequest();
  }, [requestConfig.manual, handleRequest]);

  return { data: data as T, loading, error, handleRequest };
};

export default useAxios;
