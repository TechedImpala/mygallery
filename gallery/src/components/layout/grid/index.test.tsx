import React from "react";
import { act, fireEvent, render, screen } from "@testing-library/react";
import Grid from "./";
import { useAppContext } from "../../../context/AppContext";
import { ApplicationStateEnum } from "../../../types";
import useAxios from "../../../hooks/useAxios";
import { photosResponse, topicResponse } from "../__fixtures___/index.fixtures";

jest.mock("../../../context/AppContext");
jest.mock("../../../hooks/useAxios");

describe("Grid:", () => {
  test("loads grid when a topic is selected", () => {
    const handleRequestMock = jest.fn();

    (useAxios as jest.Mock).mockReturnValue({
      data: photosResponse,
      loading: false,
      handleRequest: handleRequestMock,
    });
    (useAppContext as jest.Mock).mockReturnValue({
      activeState: ApplicationStateEnum.MENU,
      activeTopic: "color-of-water",
    });

    const { container } = render(<Grid />);

    expect(handleRequestMock).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  test("sets grid active when selected", () => {
    const handleRequestMock = jest.fn();

    (useAxios as jest.Mock).mockReturnValue({
      data: photosResponse,
      loading: false,
      handleRequest: handleRequestMock,
    });
    (useAppContext as jest.Mock).mockReturnValue({
      activeState: ApplicationStateEnum.MENU,
      activeTopic: "color-of-water",
      setActiveState: jest.fn(),
    });

    const { container, getByTestId } = render(<Grid />);

    act(() => {
      fireEvent.click(getByTestId("image-JLDWc82d3U4"));
    });

    expect(handleRequestMock).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });
});
