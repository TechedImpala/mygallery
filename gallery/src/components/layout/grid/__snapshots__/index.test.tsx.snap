// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid: loads grid when a topic is selected 1`] = `
<div>
  <div
    class="sc-hKMtZM hiNdfX"
  >
    <div
      class="sc-eCYdqJ deoAcS"
    >
      <img
        class="sc-jSMfEi jGEwmv"
        color="#c0c0c0"
        data-testid="image-JLDWc82d3U4"
        src="http://gallery.test.com/photo-1650324705564-2eb0b6b23954?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjA4OTh8MHwxfHRvcGljfHxibzhqUUtUYUUwWXx8fHx8Mnx8MTY1MDQwNDQwNQ&ixlib=rb-1.2.1&q=80&w=200"
      />
      <img
        class="sc-jSMfEi jGEwmv"
        color="#d9d9c0"
        data-testid="image-xcEgnET1zHM"
        src="http://gallery.test.com/photo-1650361072639-e2d0d4f7f3e6?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjA4OTh8MHwxfHRvcGljfHxibzhqUUtUYUUwWXx8fHx8Mnx8MTY1MDQwNDQwNQ&ixlib=rb-1.2.1&q=80&w=200"
      />
    </div>
  </div>
</div>
`;

exports[`Grid: sets grid active when selected 1`] = `
<div>
  <div
    class="sc-hKMtZM hiNdfX"
  >
    <div
      class="sc-eCYdqJ deoAcS"
    >
      <img
        class="sc-jSMfEi NjztP"
        color="#c0c0c0"
        data-testid="image-JLDWc82d3U4"
        src="http://gallery.test.com/photo-1650324705564-2eb0b6b23954?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjA4OTh8MHwxfHRvcGljfHxibzhqUUtUYUUwWXx8fHx8Mnx8MTY1MDQwNDQwNQ&ixlib=rb-1.2.1&q=80&w=200"
      />
      <img
        class="sc-jSMfEi jGEwmv"
        color="#d9d9c0"
        data-testid="image-xcEgnET1zHM"
        src="http://gallery.test.com/photo-1650361072639-e2d0d4f7f3e6?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=MnwzMjA4OTh8MHwxfHRvcGljfHxibzhqUUtUYUUwWXx8fHx8Mnx8MTY1MDQwNDQwNQ&ixlib=rb-1.2.1&q=80&w=200"
      />
    </div>
  </div>
</div>
`;
